File: homework/hw-5/python/sql_test_4.py

import psycopg as pg
import config

def main():
    # connection info
    hst = config.HOST
    usr = config.USER
    pwd = config.PASSWORD
    dat = config.DATABASE

    # make a connection
    with pg.connect(host=hst, user=usr, password=pwd, dbname=dat) as cn:
        # get a pet id, make sure it is unique
        pet_id = int(input('Please enter a pet id: '))
        q = "SELECT * FROM pet WHERE id = %s"

        # check that the id exists
        with cn.cursor() as rs:
            rs.execute(q, (pet_id,))
            if not rs.fetchone(): 
                print('This pet id is invalid')
                rs.close()
                cn.close()
                return

        # execute the update
        q = "DELETE FROM pet WHERE id = %s"
        with cn.cursor() as rs:
            # make the change
            rs.execute(q, (pet_id,))
            # make it stick
            cn.commit()

        # print the exit message
        print('Pet id', pet_id, 'has been removed from the database')

    
if __name__ == '__main__':
    main()

File: homework/hw-5/cpp/sql_test_4.cpp

#include <iostream>
#include <pqxx/pqxx>
#include "config.h"

using namespace std;
using namespace pqxx;

int main()
{
  try {
    // connection info
    string usr = config::USER;
    string pwd = config::PASSWORD;
    string hst = config::HOST;
    string dat = config::DATABASE;
    string url = "postgresql://" + usr + ":" + pwd + "@" + hst + "/" + dat;
    
    // create a connection
    connection cx{url};
    work tx(cx);

    // get a pet id, make sure it is valid
    cout << "Please enter a pet id: ";
    int id;
    cin >> id;
    cin.ignore();
    string q = "SELECT * FROM pet WHERE id = $1";
    cx.prepare("q1", q);
    result r{tx.exec_prepared("q1", id)};
    if (r.empty()) {
      cout << "This pet id is invalid" << endl;
      return 1;
    }

    q = "DELETE FROM pet WHERE id = $1";
    cx.prepare("d1", q);
    tx.exec_prepared("d1", id);

    // needed here to make the insert (database update) stick!
    tx.commit();

    cout << "Pet id " << id << " has been removed from the database" << endl;

  }    
  catch (sql_error const &e) {
    cerr << "SQL error: " << e.what() << endl;
    return 1;
  }
  catch (exception const &e) {
    cerr << "Error: " << e.what() << endl;
    return 1;
  }
  
}
